src/realm/object-store/sectioned_results.cpp

```diff
@@ -0,0 +1,419 @@
+    m_previous_str_buffers.swap(m_current_str_buffers);
+    m_previous_key_to_index_lookup.clear();
+    m_prev_section_index_to_key.clear();
+    m_ordered_section_keys.clear();
+    for (auto& [key, section] : m_sections) {
+        m_previous_key_to_index_lookup[key] = section.index;
+        m_prev_section_index_to_key[section.index] = key;
+    }
+
+    size_t size = m_results.size();
+
+    m_sections.clear();
+    m_row_to_index_path.clear();
+    m_row_to_index_path.resize(size);
+
+    for (size_t i = 0; i < size; ++i) {
+        Mixed key = m_callback(m_results.get_any(i), m_results.get_realm());
+        // Disallow links as section keys. It would be uncommon to use them to begin with
+        // and if the object acting as the key was deleted bad things would happen.
+        if (key.is_type(type_Link, type_LinkList, type_TypedLink)) {
+            throw std::logic_error("Links are not supported as section keys.");
+        }
+
+        auto it = m_sections.find(key);
+        if (it == m_sections.end()) {
+            if (!key.is_null() && key.is_type(type_String, type_Binary)) {
+                (key.get_type() == type_String) ? create_buffered_key(key, m_current_str_buffers, key.get_string())
+                                                : create_buffered_key(key, m_current_str_buffers, key.get_binary());
+            }
+
+            auto idx = m_sections.size();
+            Section section;
+            section.key = key;
+            section.index = idx;
+            section.indices.push_back(i);
+            m_sections[key] = section;
+            m_row_to_index_path[i] = {idx, section.indices.size() - 1};
+            m_ordered_section_keys.push_back(key);
+        }
+        else {
+            auto& section = it->second;
+            section.indices.push_back(i);
+            m_row_to_index_path[i] = {section.index, section.indices.size() - 1};
+        }
+    }
+}
+
+size_t SectionedResults::size()
+{
+    calculate_sections_if_required();
+    return m_sections.size();
+}
+
+ResultsSection SectionedResults::operator[](size_t idx)
+{
+    REALM_ASSERT(is_valid());
+    size_t s = size();
+    REALM_ASSERT(idx < s && s == m_ordered_section_keys.size());
+    return ResultsSection(this, m_ordered_section_keys[idx]);
+}
+
+NotificationToken SectionedResults::add_notification_callback(SectionedResultsNotificatonCallback callback,
+                                                              KeyPathArray key_path_array) &
+{
+    return m_results.add_notification_callback(SectionedResultsNotificationHandler(*this, std::move(callback)),
+                                               std::move(key_path_array));
+}
+
+NotificationToken SectionedResults::add_notification_callback_for_section(
+    Mixed section_key, SectionedResultsNotificatonCallback callback, KeyPathArray key_path_array)
+{
+    auto it = m_current_str_buffers.find(section_key);
+    REALM_ASSERT(it != m_current_str_buffers.end());
```

Comment:
This assertion is incorrect. The section key will not be present in `m_current_str_buffers` if it is anything other than a non-empty string. This also points at some missing tests, such as sectioning on an int.